### PR TITLE
Integration of multiGPU training

### PR DESCRIPTION
**Pull request by T-Almeida**
**TL;DR:** Every legacy training script will be compatible with the new multiGPU training

Integration of multiGPU (Data parallelism) training with **0** changes to the training script. 

This new mechanism leverages the [horovod ](https://github.com/horovod/horovod) library to distribute the work between processes, and it follows an MPI like API. 

Currently, to run on a multiGPU setup the [horovod.tensorflow](https://horovod.readthedocs.io/en/stable/api.html#module-horovod.tensorflow) package must be installed, if not the code will default to singleGPU.

Example of how to run on 2 GPUs using horovod:
`CUDA_VISIBLE_DEVICES="0,1" horovodrun -np 2 -H localhost:2 python tutorials/classifier_example.py`

For singleGPU just ignore the horovodrun script:
`CUDA_VISIBLE_DEVICES="0" python tutorials/classifier_example.py`

Note, that the same code is used, so by default, the code written for singleGPU should also work on multiGPU. Behind the scenes, Polus is handling the sharding of the dataset (among the processes), the distributed averaging of the gradients, the distributed initialization, the distributed execution of the validation callbacks (validation dataset is also distributed)...

### CODE REVIEW

**Comment by T-Almeida** [comment]
The default implementation was tested with horovod 0.23.0, since 0.24 seems to break during the installation...

Deeper tests are still required, especially using the NCCL (and MPI) has the communication library instead of GLOO.

**Comment by T-Almeida** [comment]
Updates about the installation of horovod: I was not able to correctly install horovod with MPI and NCCL using conda. However, I was able to lunch the horovod docker image that seems to have a correct installation of horovod with NCCL and MPI:

Output of `horovodrun --check-build`
```
Horovod v0.24.2:

Available Frameworks:
    [X] TensorFlow
    [X] PyTorch
    [X] MXNet

Available Controllers:
    [X] MPI
    [X] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [X] Gloo
```

**Comment by T-Almeida** [comment]
Our current setup uses a single k80 card (2 gpu) and was not able to perform all_reduce op by using the default P2P NCCL communication. However, when disabled with "NCCL_P2P_DISABLE=1" the GPU uses shared memory for communication, which turns out to be much more efficient than the default Gloo library (which uses CPU).

`NCCL_P2P_DISABLE=1 NCCL_DEBUG=INFO horovodrun -np 2 -H localhost:2 python tutorials/classifier_example.py`
